src/apps/polynomial/optimism/polynomial.balance-fetcher.ts

```diff
@@ -20,29 +22,88 @@ export class OptimismPolynomialBalanceFetcher implements BalanceFetcher {
   ) {}
 
-  async getVaultBalances(address: string) {
+  private async getPoolTokenBalances(address: string) {
+    return this.appToolkit.helpers.tokenBalanceHelper.getTokenBalances({
+      address,
+      appId,
+      groupId: POLYNOMIAL_DEFINITION.groups.vaults.id,
+      network,
+    });
+  }
+
+  async getPendingDepositBalances(address: string) {
     return this.appToolkit.helpers.contractPositionBalanceHelper.getContractPositionBalances({
       address,
       appId: POLYNOMIAL_DEFINITION.id,
       groupId: POLYNOMIAL_DEFINITION.groups.vaults.id,
       network,
-      resolveBalances: async ({ address, network, multicall, contractPosition: position }) => {
-        const token = position.tokens[0];
-        const contract = this.contractFactory.vaults({ address: resolverAddress, network });
-        const balances = await multicall.wrap(contract).getUserBalance(address, position.address);
-        const totalBalance =
-          Number(balances._balance) + Number(balances._withdrawToComplete) + Number(balances._cancellableWithdraw);
-        return [drillBalance(token, totalBalance.toString())];
+      resolveBalances: async ({ address, multicall, contractPosition: position }) => {
+        const contract = this.contractFactory.polynomialCoveredCall(position);
+
+        const depositHead = Number(await multicall.wrap(contract).queuedDepositHead());
+        // Note: ignores pending deposits when deposit queue is large. Consider checking if the last user is ZERO_ADDRESS and recursing
+        const pendingDeposits = await Promise.all(
+          map(range(250), async i => multicall.wrap(contract).depositQueue(depositHead + Number(i))),
+        );
+        const pendingDepositBalance = sumBy(pendingDeposits, deposit => {
+          // Note: ignores partial deposits
+          if (deposit.user.toLowerCase() === address.toLowerCase() && !Number(deposit.mintedTokens)) {
+            return Number(deposit.depositedAmount);
+          }
+          return 0;
+        });
+
+        return [drillBalance(position.tokens.find(isSupplied)!, pendingDepositBalance.toString())];
+      },
+    });
+  }
+
+  async getPendingWithdrawalBalances(address: string) {
+    return this.appToolkit.helpers.contractPositionBalanceHelper.getContractPositionBalances({
+      address,
+      appId: POLYNOMIAL_DEFINITION.id,
+      groupId: POLYNOMIAL_DEFINITION.groups.vaults.id,
+      network,
+      resolveBalances: async ({ address, multicall, contractPosition: position }) => {
+        const contract = this.contractFactory.polynomialCoveredCall(position);
+
+        const withdrawalHead = Number(await multicall.wrap(contract).queuedWithdrawalHead());
+        // Note: ignores pending withdrawals when withdrawal queue is large
+        const pendingWithdrawals = await Promise.all(
+          map(range(250), async i => multicall.wrap(contract).withdrawalQueue(withdrawalHead + Number(i))),
+        );
```

Comment:
I'm pretty sure you can use `nextQueuedWithdrawalId` with `queuedWithdrawalHead`:

```
const start = Number(await multicall.wrap(contract).queuedWithdrawalHead());
const end = Number(await multicall.wrap(contract).nextQueuedWithdrawalId());
const pendingWithdrawals = await Promise.all(range(start, end).map(i => multicall.wrap(contract).depositQueue(i)));
```

(`nextQueuedWithdrawalId` is the next "free slot")